proxy/src/main/java/com/velocitypowered/proxy/connection/registry/DimensionRegistry.java

```diff
@@ -0,0 +1,128 @@
+package com.velocitypowered.proxy.connection.registry;
+
+import java.util.HashSet;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import javax.annotation.Nonnull;
+
+import net.kyori.nbt.CompoundTag;
+import net.kyori.nbt.ListTag;
+import net.kyori.nbt.Tag;
+import net.kyori.nbt.TagType;
+
+public class DimensionRegistry {
+
+  private final @Nonnull Set<DimensionData> dimensionRegistry;
+  private final @Nonnull String[] levelNames;
+
+  /**
+   * Initializes a new {@link DimensionRegistry} instance.
+   * This registry is required for 1.16+ clients/servers to communicate,
+   * it constrains the dimension types and names the client can be sent
+   * in a Respawn action (dimension change).
+   * @param dimensionRegistry a populated set containing dimension data types
+   * @param levelNames a populated {@link Set} of the dimension level names the server offers
+   */
+  public DimensionRegistry(Set<DimensionData> dimensionRegistry,
+                           String[] levelNames) {
+    if (dimensionRegistry == null || dimensionRegistry.isEmpty()
+            || levelNames == null || levelNames.length == 0) {
+      throw new IllegalArgumentException(
+              "Dimension registry requires valid arguments, not null and not empty");
+    }
+    this.dimensionRegistry = dimensionRegistry;
+    this.levelNames = levelNames;
+  }
+
+  public @Nonnull Set<DimensionData> getDimensionRegistry() {
+    return dimensionRegistry;
+  }
+
+  public @Nonnull String[] getLevelNames() {
+    return levelNames;
+  }
+
+  /**
+   * Returns the internal dimension data type as used by the game.
+   * @param dimensionIdentifier how the dimension is identified by the connection
+   * @return game dimension data
+   */
+  public @Nonnull DimensionData getDimensionData(@Nonnull String dimensionIdentifier) {
+    if (dimensionIdentifier == null) {
+      throw new IllegalArgumentException("Dimension identifier cannot be null!");
+    }
+    for (DimensionData iter : dimensionRegistry) {
```

Comment:
This won't hold up merging this in, but perhaps we can consider using a map of dimension IDs to dimension data (try Guava `Maps.uniqueIndex`).

Reply:
I had that but I dropped it because I felt like it’s a waste of space seeing as the DimensionData object already needs to store that